### PR TITLE
Add a CSI driver for Alluxio

### PR DESCRIPTION
**Pull request by mingfang**
This PR ports changes from https://github.com/mingfang/alluxio-csi.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,22 @@
+module github.com/mingfang/alluxio-csi
```
Comment (Xenorith) [inline go.mod]: can you update this to be `module github.com/Alluxio/alluxio-csi`?

**Review by Xenorith — CHANGES_REQUESTED** [review]

File: go.mod (hunk 1 above)
Comment (mingfang) [reply go.mod]: done

**Review by Xenorith — APPROVED** [review]
looks good to me, thanks

File: alluxio/controllerserver.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+func sanitizeVolumeID(volumeID string) string {
+	volumeID = strings.ToLower(volumeID)
+	if len(volumeID) > 63 {
+		h := sha1.New()
+		io.WriteString(h, volumeID)
+		volumeID = hex.EncodeToString(h.Sum(nil))
+	}
+	return volumeID
+}
```
Comment (jiacheliu3) [inline alluxio/controllerserver.go]: Please add a newline at the end of file.

File: alluxio/driver.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+func (d *driver) Run() {
+    s := csicommon.NewNonBlockingGRPCServer()
+    s.Start(
+        d.endpoint,
+        csicommon.NewDefaultIdentityServer(d.csiDriver),
+        d.newControllerServer(),
+        d.newNodeServer(),
+    )
+    s.Wait()
+}
```
Comment (jiacheliu3) [inline alluxio/driver.go]: Newline

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+	github.com/spf13/cobra v0.0.3
+	github.com/spf13/pflag v1.0.3 // indirect
+	golang.org/x/net v0.0.0-20190514140710-3ec191127204
+	golang.org/x/sys v0.0.0-20190516110030-61b9204099cb // indirect
+	golang.org/x/text v0.3.2 // indirect
+	google.golang.org/appengine v1.4.0 // indirect
+	google.golang.org/genproto v0.0.0-20190516172635-bb713bdc0e52 // indirect
+	google.golang.org/grpc v1.20.1
+	k8s.io/apimachinery v0.0.0-20190515023456-b74e4c97951f // indirect
+	k8s.io/kubernetes v1.14.2
```
Comment (jiacheliu3) [inline go.mod:20]: @madanadit Should we stick to K8s v1.16 for CSI as well?

File: main.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+				cpid, _ := syscall.Wait4(-1, nil, syscall.WNOHANG, nil)
+				if cpid < 1 {
+					break
+				}
+
+				glog.V(4).Infof("Reaped process with pid %d", cpid)
+			}
+		}
+	}()
+}
```
Comment (jiacheliu3) [inline main.go:85]: Newline

File: alluxio/controllerserver.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+import (
+    "crypto/sha1"
+    "encoding/hex"
+    "github.com/container-storage-interface/spec/lib/go/csi"
+    "github.com/golang/glog"
+    csicommon "github.com/kubernetes-csi/drivers/pkg/csi-common"
+    "golang.org/x/net/context"
+    "google.golang.org/grpc/codes"
+    "google.golang.org/grpc/status"
+    "io"
```
Comment (jiacheliu3) [inline alluxio/controllerserver.go]: nit: Please separate the imports. I think the format tool (We use `godelw format` but not sure what you typically refer to) will put `"io"` and `"strings"` at the top and have an extra row before the rest of imports. Please ignore this if that's what your format tool returns. @Xenorith has more say on this.

**Review by jiacheliu3 — CHANGES_REQUESTED** [review]

File: alluxio/controllerserver.go (hunk 6 above)
Comment (Xenorith) [reply alluxio/controllerserver.go]: more importantly the imports across each file should follow the same pattern. this one organizes all of them by alphabetical order but the subsequent ones group native imports first, then vendored imports. i have a slight preference for grouping but i'm happy as long as there's consistency across files.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/golang/glog"
+	"github.com/mingfang/alluxio-csi/alluxio"
```
Comment (madanadit) [inline main.go:11]: same here

File: go.mod (hunk 4 above)
Comment (madanadit) [reply go.mod:20]: @mingfang is this the minimum version required? can the same csi artifact be deployed to multiple version of k8s?

**Review by madanadit — CHANGES_REQUESTED** [review]

**Comment by mingfang** [comment]
The go.mod content was populated by `go mod tidy` and have been tested to work against Kubernetes v1.16

**Comment by madanadit** [comment]
@mingfang can you PTAL at the minor comments remaining?

**Comment by madanadit** [comment]
ping @mingfang

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+RUN mkdir -p $GOPATH/src/github.com/Alluxio/alluxio-csi
+COPY . $GOPATH/src/github.com/Alluxio/alluxio-csi
+
+FROM dev as build
+
+RUN cd $GOPATH/src/github.com/Alluxio/alluxio-csi && \
+    CGO_ENABLED=0 go build -o /usr/local/bin/alluxio-csi
+
+FROM alluxio/alluxio-fuse:2.0.1 as final
+COPY --from=build /usr/local/bin/alluxio-csi /usr/local/bin/
```
Comment (jiacheliu3) [inline Dockerfile:14]: Missing a newline here

Comment (mingfang) [reply Dockerfile:14]: go ahead and format it yourself.

**Pull request merged**